File: Java/004.java
public class Problem004 {

    public static boolean isPalindrome(int num) {
        String s = String.valueOf(num);
        int n = s.length();
        for (int i = 0; i < n / 2; i++) {
            if (s.charAt(i) != s.charAt(n - 1 - i)) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        int res = 0;
        for (int x = 999; x > 100; x--) {
            for (int y = 999; y > 100; y--) {
                int product = x * y;
                if (product < res) {
                    break;
                }
                if (isPalindrome(product)) {
                    res = product;
                }
            }
        }
        System.out.println("Largest palindrome: " + res);
    }
}

File: Python/004.py
# A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 x 99.
# Find the largest palindrome made from the product of two 3-digit numbers.

import math


def is_palindrome(num):
    """Check if a num is a palindrome."""
    s = str(num)
    for idx in range(len(s) // 2):
        if s[idx] != s[len(s) - 1 - idx]:
            return False
    return True


if __name__ == "__main__":
    res = -math.inf
    for x in range(999, 100, -1):
        for y in range(999, 100, -1):
            if x * y < res:
                break
            if is_palindrome(x * y):
                res = x * y
    print("Largest palindrome", res)
